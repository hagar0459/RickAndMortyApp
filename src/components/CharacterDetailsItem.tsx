/**
 * @flow
 * Created by Hagar Abdelghafar on 04.07.2020
 */
import React, { FC } from 'react';
import { Image, StyleSheet, Text, View } from 'react-native';
import type { CharcterDetailsType } from '../graphql/queries/requests';


type Props = {
  info: CharcterDetailsType | undefined;
};


export const CharacterDetailsItem: FC<Props> = ({ info }: Props) => {

  const renderInfoContainer = () => (
    <View style={styles.infoContainer}>
      <Text style={styles.name}>{info?.character.name}</Text>
      <Text style={styles.gender}>{info?.character.gender}</Text>


      <Text style={styles.species}>{info?.character.species}</Text>

    </View>
  );
  const renderImage = () => {
    return (
      <Image
        style={styles.image}
        source={{ uri: info?.character.image }}
        resizeMode={'cover'}
      />
    );
  };

  const renderContent = () => {

    return (
      <>
        {renderImage()}
        {renderInfoContainer()}

      </>
    );


  };


  return (
    <View style={styles.card} >
      {renderContent()}
    </View>
  );
}

const styles = StyleSheet.create({

  card: {
    borderRadius: 0,
    marginBottom: 20

  },
  image: {
    height: 150,

  },
  infoContainer: {
    padding: 16,
  },
  name: {
    fontSize: 22,
    fontWeight: 'bold',
  },
  gender: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 8,
  },
  species: {
    fontSize: 16,
    fontWeight: '400',
    color: '#787878',
  },
});
