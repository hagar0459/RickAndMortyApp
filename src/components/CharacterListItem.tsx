/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import React, { FC } from 'react';
import { Dimensions, Image, StyleSheet, Text, View, TouchableOpacity } from 'react-native';
import type { CharcterResultType } from '../graphql/queries/requests';


type Props = {
  info: CharcterResultType;
  onPress(item: CharcterResultType): void;
  index: number;
};
const ITEM_HEIGHT = 80;

export const CharacterListItem: FC<Props> = ({ info, onPress, index }: Props) => {

  const renderText = () => (
    <View  >
      <Text style={styles.title}>{info.name}</Text>
    </View>
  );
  const renderImage = (isRight: boolean) => {
    const rawUri = info.image;
    if (!rawUri) {
      return null;
    }
    const uri = `${rawUri ?? ''}?width=300`;
    // console.log({ uri });
    return (
      <View style={isRight ? styles.right : styles.left}>
        <Image source={{ uri }} resizeMode="cover" style={styles.image} />
      </View>
    );
  };

  const renderContent = () => {
    const isRight = index % 2 === 0;
    if (isRight) {
      return (
        <>
          {renderText()}
          {renderImage(isRight)}
        </>
      );
    }
    return (
      <>
        {renderImage(isRight)}
        {renderText()}
      </>
    );
  };


  return (
    <TouchableOpacity style={styles.card}

      onPress={() => {
        onPress(info);
      }}
    >

      {renderContent()}
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  card: {
    backgroundColor: 'white',
    borderRadius: 16,
    shadowOpacity: 0.2,
    shadowRadius: 4,
    shadowColor: 'black',
    shadowOffset: {
      height: 0,
      width: 0,
    },
    elevation: 1,
    marginVertical: 10,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',


  },
  image: {
    height: ITEM_HEIGHT,
    width: ITEM_HEIGHT,
  },
  left: {
    borderTopLeftRadius: 8,
    borderBottomLeftRadius: 8,
    overflow: 'hidden',

  },
  right: {
    borderTopRightRadius: 8,
    borderBottomRightRadius: 8,
    overflow: 'hidden',

  },
  title: {
    fontSize: 17,
    padding: 5,
    width: Dimensions.get('window').width - ITEM_HEIGHT - 30,
    textAlign: 'center',
  },
});
