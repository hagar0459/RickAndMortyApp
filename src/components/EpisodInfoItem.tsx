/**
 * @flow
 * Created by Hagar Abdelghafar on 05.06.2021
 */
import React, { FC } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import type { EpisodeType } from '../graphql/queries/requests';


type Props = {
  info: EpisodeType;

};

export const EpisodInfoItem: FC<Props> = ({ info }: Props) => {

  return (
    <View style={styles.card}  >
      <Text style={styles.name}>{info.name}</Text>
      <Text style={styles.air_date}>{info.air_date}</Text>
    </View>



  );
}

const styles = StyleSheet.create({
  card: {
    backgroundColor: 'white', margin: '5%', borderRadius: 10, width: '40%', padding: 10, alignContent: 'center', alignItems: 'center'
  },
  name: {
    fontSize: 15,
    fontWeight: '400',
    marginBottom: 8,
  },
  air_date: {
    fontSize: 12,
    fontWeight: '200',
    color: '#787878',
  },
});
