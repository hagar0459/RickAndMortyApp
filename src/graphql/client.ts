/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import { ApolloClient, InMemoryCache } from "@apollo/client";
import { GRAPHQL_URL } from "../config";

export const apolloClient = new ApolloClient({
  uri: GRAPHQL_URL,
  cache: new InMemoryCache({}),
  defaultOptions: {
    watchQuery: {
      fetchPolicy: "cache-first",
    },
  },
});
