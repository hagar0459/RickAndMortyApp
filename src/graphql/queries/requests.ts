/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import { gql } from '@apollo/client';
export const SEARCH_CHARACTER = gql`
  query SearchCharacter($page: Int!, $name: String!) {
    characters(page: $page, filter: { name: $name }) {
      results {
        name
        image
        id
      }
      info {
        next
        count
      }
    }
  }
`;
export const GET_CHARACTER_DETAILS = gql`
  query GetCharacterDetails($id: ID!) {
    character(id: $id) {
      name
      id
      image
      gender
      species
      episode {
        air_date
        name
      }
    }
  }
`;
export type CharcterListType = {
  characters: {
    info: characterListInfoType;
    results: Array<CharcterResultType>;
  };
};
export type CharcterResultType = {
  id: number;
  name: string;
  image: string;
};
export type characterListInfoType = {
  next: number | null;
  count: number;
};
export type CharcterDetailsType = {
  character: {
    id: number;
    name: string;
    image: string;
    gender: string;
    species: string;
    episode: Array<EpisodeType>;
  };
};
export type EpisodeType = {
  air_date: string;
  name: string;
};
