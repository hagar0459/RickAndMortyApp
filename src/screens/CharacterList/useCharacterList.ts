/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import React, { useState, useEffect } from "react";
import { useLazyQuery } from "@apollo/client";
import {
  SEARCH_CHARACTER,
  CharcterListType,
  CharcterResultType,
} from "../../graphql/queries/requests";
import { Alert } from "react-native";

type Props = {
  searchTxt: String;
  pageNumber: number;
};

type Result = {
  characters: Array<CharcterResultType>;
  getCharactersList(): void;
  loading: boolean;
  loadMore(): void;
  error: string | undefined;
};

export const useCharacterList = ({ searchTxt, pageNumber }: Props): Result => {
  const [characters, setcharacters] = useState<CharcterResultType[]>([]);
  const [currentPage, setCurrentPage] = useState(1);

  const [getCharactersList, { data, error, loading }] =
    useLazyQuery<CharcterListType>(SEARCH_CHARACTER, {
      variables: {
        page: currentPage,
        name: searchTxt,
      },
    });

  useEffect(() => {
    if (!loading) {
      getCharactersList();
    }
  }, [currentPage, getCharactersList]); 

  useEffect(() => {
    if (data?.characters?.results && currentPage === 1) {
      setcharacters(data?.characters?.results);
    } else if (
      data?.characters?.results &&
      data?.characters.info.next !== null
    ) {
      setcharacters([...characters, ...data?.characters?.results]);
    }
  }, [data]);

  useEffect(() => {
    setCurrentPage(pageNumber);
  
  }, [searchTxt]);

  const loadMore = () => {
    if (loading) {
      return;
    } else if (data) {
      if (characters.length < data?.characters?.info?.count) {
        setCurrentPage(currentPage + 1);
      }
    }
  };

  return {
    characters,
    getCharactersList,
    loading,
    loadMore,
    error: error?.message,
  };
};
