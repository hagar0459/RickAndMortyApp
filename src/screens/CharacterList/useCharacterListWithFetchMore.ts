/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import { useState, useEffect } from 'react';
import { useLazyQuery } from '@apollo/client';
import {
  SEARCH_CHARACTER,
  CharcterListType,
  CharcterResultType,
} from '../../graphql/queries/requests';

type Props = {
  searchTxt: String;
  pageNumber: number;
};

type Result = {
  characters: Array<CharcterResultType>;
  getCharactersList(): void;
  loading: boolean;
  loadMore(): void;
  error: string | undefined;
};

export const useCharacterListWithFetchMore = ({
  searchTxt,
  pageNumber,
}: Props): Result => {
  const [characters, setcharacters] = useState<CharcterResultType[]>([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [getCharactersList, { data, error, loading, fetchMore }] =
    useLazyQuery<CharcterListType>(SEARCH_CHARACTER, {
      variables: {
        page: currentPage,
        name: searchTxt,
      },
    });

  useEffect(() => {
    if (!loading) {
      getCharactersList();
      // return () => {
      //   unsubscribe();
      // };
    }
  }, [currentPage, getCharactersList]);

  useEffect(() => {
    if (data?.characters?.results && currentPage === 1) {
      setcharacters(data?.characters?.results);
    } else if (
      data?.characters?.results &&
      data?.characters.info.next !== null
    ) {
      setcharacters([...characters, ...data?.characters?.results]);
    }
  }, [data]);

  useEffect(() => {
    setCurrentPage(pageNumber);
  }, [searchTxt]);

  const loadMore = () => {
    pageNumber++;
    debugger;
    if (data?.characters?.info?.next)
      fetchMore({
        variables: {
          page: pageNumber,
          name: searchTxt,
        },
        updateQuery: (prev, { fetchMoreResult }) => {
          if (!fetchMoreResult) return prev;
          const subscriptionResponse = fetchMoreResult.characters.results;

          const newCache = Object.assign({}, prev, {
            characters: {
              results: [subscriptionResponse, ...prev.characters.results],
            },
          });

          return newCache;

          // if (!fetchMoreResult) return prev;
          // return (fetchMoreResult.characters.results = [
          //   ...prev.characters.results,
          //   ...fetchMoreResult.characters.results,
          // ]);
        },
      });
  };

  return {
    characters,
    getCharactersList,
    loading,
    loadMore,
    error: error?.message,
  };
};
