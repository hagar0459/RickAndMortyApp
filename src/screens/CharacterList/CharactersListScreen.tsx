/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */

import React, { FC, useState, useRef } from 'react';
import
{
  FlatList,
  StyleSheet,
  ActivityIndicator,
  SafeAreaView,
  TextInput,
  View,
  ListRenderItemInfo,
  Alert,
  Text,
} from 'react-native';
import { CharcterResultType } from '../../graphql/queries/requests';
import { CharacterListItem } from '../../components/CharacterListItem';
import { NAVIGATION_CHARACTERS_DETAILS_ROUTE } from '../../Navigation';
import { useNavigation } from '@react-navigation/native';
import { useCharacterList } from './useCharacterList';

export const CharactersListScreen: FC = () =>
{
  let ref = useRef<FlatList<CharcterResultType>>( null );

  const [searchTxt, setsearchTxt] = useState( '' );
  const { loading, characters, loadMore, error } = useCharacterList( {
    searchTxt: searchTxt,
    pageNumber: 1,
  } );
  const navigation = useNavigation();
  if ( error )
  {
    if ( !characters || error.includes( 'Network' ) )
    {
      Alert.alert( error );
    }
  }
  const footerComponent = () =>
  {
    if ( loading && characters.length !== 0 )
    {
      return (
        <View style={{ height: 20 }}>
          <ActivityIndicator size='large' />
        </View>
      );
    }
    return null;
  };
  const listEmptyComponent = () =>
  {
    if ( !loading && characters.length === 0 )
    {
      return (
        <View
          style={{
            justifyContent: 'center',
            alignItems: 'center',
            width: '100%',
          }}>
          {error?.includes( 'Network' ) && (
            <Text style={{ textAlign: 'center' }}> Network Error!!!</Text>
          )}
          {!error && (
            <Text style={{ textAlign: 'center' }}>
              {' '}
              No Characters Found !!!
            </Text>
          )}
        </View>
      );
    }
    return null;
  };
  const renderCharacter = ( {
    item,
    index,
  }: ListRenderItemInfo<CharcterResultType> ) =>
  {
    return (
      <CharacterListItem
        info={item}
        onPress={() =>
        {
          navigation.navigate( NAVIGATION_CHARACTERS_DETAILS_ROUTE, {
            characterId: item.id,
            title: item.name,
          } );
        }}
        index={index}
      />
    );
  };
  return (
    <SafeAreaView>
      <View style={{ height: 60, width: '100%', backgroundColor: 'white' }}>
        <TextInput
          style={styles.input}
          placeholder='search...'
          onChangeText={( text ) =>
          {

            ref.current?.scrollToOffset( { animated: true, offset: 0 } );
            setsearchTxt( text );
          }}
          autoCorrect={false}
          value={searchTxt}
        />
      </View>
      <FlatList
        keyboardShouldPersistTaps='never'
        ref={ref}
        style={styles.charctersList}
        contentContainerStyle={styles.charctersListContainer}
        data={characters}
        keyExtractor={item => `${ item.id }`}
        renderItem={renderCharacter}
        onEndReached={loadMore}
        ListFooterComponent={footerComponent}
        ListEmptyComponent={listEmptyComponent}
      />
    </SafeAreaView>
  );
};
const styles = StyleSheet.create( {
  charctersList: {
    backgroundColor: '#fafafa',
  },
  charctersListContainer: {
    backgroundColor: '#fafafa',
    paddingVertical: 8,
    marginHorizontal: 8,
  },
  input: {
    flex: 1,
    marginTop: 10,
    fontSize: 18,
    paddingHorizontal: 10,
    backgroundColor: '#dcdcdc',
    borderRadius: 10,
    marginBottom: 10,
    width: '90%',
    marginHorizontal: '5%',
  },
} );
