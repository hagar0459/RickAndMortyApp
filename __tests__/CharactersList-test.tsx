import * as React from 'react';
import { CharactersListScreen } from '../src/screens/CharacterList/CharactersListScreen';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import {MockedProvider} from '@apollo/client/testing';
import {SEARCH_CHARACTER} from '../src/graphql/queries/requests';
import { FlatList ,TextInput} from 'react-native';
import{CharacterListItem} from '../src/components/CharacterListItem'

async function wait(ms = 0) {
  await renderer.act(() => {
    return new Promise(resolve => {
      setTimeout(resolve, ms);
    });
  });
}

describe('Testing CharactersList', () => {


it('Screen Shows FlatList', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharactersListScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(FlatList)).not.toThrow();
  });

it('Screen Shows InputText', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharactersListScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(TextInput)).not.toThrow();
  });

it('Screen Shows CharcterListItem', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharactersListScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(CharacterListItem)).not.toThrow();
  });

it('Shows Characters List ', async () => {
    const mockedData = {
      request: {
        query: SEARCH_CHARACTER,
        variables: { 
          page: 1,
          name: ''}
      },
      result: {
        "data": {
          "characters": {
            "results": [
              {
                "name": "Rick Sanchez",
                "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
                "id": "1"                
              },
              {
                "name": "Morty Smith",
                "image": "https://rickandmortyapi.com/api/character/avatar/2.jpeg",
                "id": "2"
              },
              {
                "name": "Summer Smith",
                "image": "https://rickandmortyapi.com/api/character/avatar/3.jpeg",
                "id": "3"
              },
              {
                "name": "Beth Smith",
                "image": "https://rickandmortyapi.com/api/character/avatar/4.jpeg",
                "id": "4"
              },
              {
                "name": "Jerry Smith",
                "image": "https://rickandmortyapi.com/api/character/avatar/5.jpeg",
                "id": "5"
              },
              {
                "name": "Abadango Cluster Princess",
                "image": "https://rickandmortyapi.com/api/character/avatar/6.jpeg",
                "id": "6"
              },
              {
                "name": "Abradolf Lincler",
                "image": "https://rickandmortyapi.com/api/character/avatar/7.jpeg",
                "id": "7"
              },
              {
                "name": "Adjudicator Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/8.jpeg",
                "id": "8"
              },
              {
                "name": "Agency Director",
                "image": "https://rickandmortyapi.com/api/character/avatar/9.jpeg",
                "id": "9"
              },
              {
                "name": "Alan Rails",
                "image": "https://rickandmortyapi.com/api/character/avatar/10.jpeg",
                "id": "10"
              },
              {
                "name": "Albert Einstein",
                "image": "https://rickandmortyapi.com/api/character/avatar/11.jpeg",
                "id": "11"
              },
              {
                "name": "Alexander",
                "image": "https://rickandmortyapi.com/api/character/avatar/12.jpeg",
                "id": "12"
              },
              {
                "name": "Alien Googah",
                "image": "https://rickandmortyapi.com/api/character/avatar/13.jpeg",
                "id": "13"
              },
              {
                "name": "Alien Morty",
                "image": "https://rickandmortyapi.com/api/character/avatar/14.jpeg",
                "id": "14"
              },
              {
                "name": "Alien Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/15.jpeg",
                "id": "15"
              },
              {
                "name": "Amish Cyborg",
                "image": "https://rickandmortyapi.com/api/character/avatar/16.jpeg",
                "id": "16"
              },
              {
                "name": "Annie",
                "image": "https://rickandmortyapi.com/api/character/avatar/17.jpeg",
                "id": "17"
              },
              {
                "name": "Antenna Morty",
                "image": "https://rickandmortyapi.com/api/character/avatar/18.jpeg",
                "id": "18"
              },
              {
                "name": "Antenna Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/19.jpeg",
                "id": "19"
              },
              {
                "name": "Ants in my Eyes Johnson",
                "image": "https://rickandmortyapi.com/api/character/avatar/20.jpeg",
                "id": "20"
              }
            ],
            "info": {
              "next": 2,
              "count": 671
            }
          }
        },
      },
    };
    const component = renderer.create(
      <MockedProvider mocks={[mockedData]} addTypename={false} >
        <CharactersListScreen />
      </MockedProvider>,
    );

    await wait();

    console.log(component.toJSON());
    expect(() => {
      component.root.findByProps({
        children: 'Rick Sanchez',
      });
    });
  });
  it('Shows Characters List with Serach Option', async () => {
    const mockedData = {
      request: {
        query: SEARCH_CHARACTER,
        variables: { 
          page: 1,
          name: 'Rick'}
      },
      result: {
        "data": {
          "characters": {
            "results": [
              {
                "name": "Rick Sanchez",
                "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
                "id": "1"
              },
              {
                "name": "Adjudicator Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/8.jpeg",
                "id": "8"
              },
              {
                "name": "Alien Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/15.jpeg",
                "id": "15"
              },
              {
                "name": "Antenna Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/19.jpeg",
                "id": "19"
              },
              {
                "name": "Aqua Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/22.jpeg",
                "id": "22"
              },
              {
                "name": "Black Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/48.jpeg",
                "id": "48"
              },
              {
                "name": "Bootleg Portal Chemist Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/56.jpeg",
                "id": "56"
              },
              {
                "name": "Commander Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/69.jpeg",
                "id": "69"
              },
              {
                "name": "Cool Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/72.jpeg",
                "id": "72"
              },
              {
                "name": "Cop Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/74.jpeg",
                "id": "74"
              },
              {
                "name": "Cowboy Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/78.jpeg",
                "id": "78"
              },
              {
                "name": "Cronenberg Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/82.jpeg",
                "id": "82"
              },
              {
                "name": "Cyclops Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/86.jpeg",
                "id": "86"
              },
              {
                "name": "Doofus Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/103.jpeg",
                "id": "103"
              },
              {
                "name": "Evil Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/119.jpeg",
                "id": "119"
              },
              {
                "name": "Garment District Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/135.jpeg",
                "id": "135"
              },
              {
                "name": "Insurance Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/164.jpeg",
                "id": "164"
              },
              {
                "name": "Investigator Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/165.jpeg",
                "id": "165"
              },
              {
                "name": "Juggling Rick",
                "image": "https://rickandmortyapi.com/api/character/avatar/187.jpeg",
                "id": "187"
              },
              {
                "name": "Maximums Rickimus",
                "image": "https://rickandmortyapi.com/api/character/avatar/215.jpeg",
                "id": "215"
              }
            ],
            "info": {
              "next": 2,
              "count": 84
            }
          }
        },
      },
    };
    const component = renderer.create(
      <MockedProvider mocks={[mockedData]} addTypename={false} >
        <CharactersListScreen />
      </MockedProvider>,
    );

    await wait();

    console.log(component.toJSON());
    expect(() => {
      component.root.findByProps({
        children: 'Aqua Rick',
      });
    });
  });

it('Screen Have TextInput With Place Holder', async () => {
    const component = (
        <CharactersListScreen />
    );

    const { findByPlaceholderText } = render(component);

    const header = await findByPlaceholderText('search...');

    expect(header).toBeTruthy();
  });

it("Search For A Character When The Character Name is Typed", () => {
    const { getByPlaceholderText,getAllByText } = render(<CharactersListScreen />);
    fireEvent.changeText(getByPlaceholderText("search..."), "rick");
    expect(getAllByText("rick")).toHaveLength(1);
  }); 

//pagination (onEndReach )
//errors handling 
});