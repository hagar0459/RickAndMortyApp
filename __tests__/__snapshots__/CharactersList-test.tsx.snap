// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CharactersList Link renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "height": 60,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      onChangeText={[Function]}
      placeholder="search..."
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#dcdcdc",
          "borderRadius": 10,
          "flex": 1,
          "fontSize": 18,
          "marginBottom": 10,
          "marginHorizontal": "5%",
          "marginTop": 10,
          "paddingHorizontal": 10,
          "width": "90%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    contentContainerStyle={
      Object {
        "backgroundColor": "#fafafa",
        "marginHorizontal": 8,
        "paddingVertical": 8,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="never"
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#fafafa",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
           
          No Characters Found !!!
        </Text>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`Testing CharactersList Snapshot Test 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "height": 60,
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      onChangeText={[Function]}
      placeholder="search..."
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "#dcdcdc",
          "borderRadius": 10,
          "flex": 1,
          "fontSize": 18,
          "marginBottom": 10,
          "marginHorizontal": "5%",
          "marginTop": 10,
          "paddingHorizontal": 10,
          "width": "90%",
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    contentContainerStyle={
      Object {
        "backgroundColor": "#fafafa",
        "marginHorizontal": 8,
        "paddingVertical": 8,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    keyboardShouldPersistTaps="never"
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#fafafa",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
           
          No Characters Found !!!
        </Text>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
